clang/runtime/dpct-rt/include/dpct/dpl_extras/dpcpp_extensions.h

```diff
@@ -566,6 +566,116 @@ class radix_sort {
+          load_algorithm ALGORITHM,
+          typename InputT,
+          typename InputIteratorT>
+class workgroup_load {
+public:
+  
+  __dpct_inline__ void load_blocked(size_t linear_tid, InputIteratorT block_itr,
+                                    InputT (&items)[ITEMS_PER_WORK_ITEM]) {
+  
+  uint32_t workgroup_offset = linear_tid * ITEMS_PER_WORK_ITEM;
+  if (workgroup_offset > GROUP_WORK_ITEMS){
+      uint32_t final_idx = (GROUP_WORK_ITEMS - workgroup_offset);
+      for (uint32_t idx = 0; idx < final_idx ; idx++) {
+        items[idx] = block_itr[workgroup_offset + idx];
+      }
+    }
+    else{
+    #pragma unroll
+      for (uint32_t idx = 0; idx < ITEMS_PER_WORK_ITEM; idx++) {
+        items[idx] = block_itr[workgroup_offset + idx];
+      }
+    }
+  }
+
+private:
+};
+
+template <size_t GROUP_WORK_ITEMS,
+          size_t ITEMS_PER_WORK_ITEM,
+          load_algorithm ALGORITHM,
+          typename InputT,
+          typename InputIteratorT,
+          typename Item>
+class subgroup_load {
+public:
+  __dpct_inline__ void load_striped(size_t linear_tid, InputIteratorT block_itr,
+                                    InputT (&items)[ITEMS_PER_WORK_ITEM]) {
+
+    if (linear_tid + ((ITEMS_PER_WORK_ITEM - 1)*ITEMS_PER_WORK_ITEM) > GROUP_WORK_ITEMS){
+      uint32_t final_idx = (GROUP_WORK_ITEMS - linear_tid)/ITEMS_PER_WORK_ITEM;
+      for (uint32_t idx = 0; idx < final_idx ; idx++) {
+        items[idx] = block_itr[linear_tid + (idx  * ITEMS_PER_WORK_ITEM)];
+      }
+    }
+    else{
+    #pragma unroll
+      for (uint32_t idx = 0; idx < ITEMS_PER_WORK_ITEM; idx++) {
+        items[idx] = block_itr[linear_tid + (idx  * GROUP_WORK_ITEMS)];
+      }
+    }
+  }
+
+  
+  __dpct_inline__ void load_subgroup_striped(const Item &item, size_t linear_tid,
+                                             InputIteratorT block_itr,
+                                             InputT (&items)[ITEMS_PER_WORK_ITEM]) {
+  
+    size_t subgroup_offset = item.get_sub_group().get_local_range()[0];
+    if (linear_tid + ((ITEMS_PER_WORK_ITEM - 1)*ITEMS_PER_WORK_ITEM) > GROUP_WORK_ITEMS){
+      uint32_t final_idx = (GROUP_WORK_ITEMS - linear_tid)/ITEMS_PER_WORK_ITEM;
+      for (uint32_t idx = 0; idx < final_idx ; idx++) {
+        new (&items[idx])
+          InputT(block_itr[subgroup_offset + linear_tid + (idx * ITEMS_PER_WORK_ITEM)]);
+      }
+    }
+    else{
+    #pragma unroll
+      for (uint32_t idx = 0; idx < ITEMS_PER_WORK_ITEM; idx++) {
+        new (&items[idx])
+          InputT(block_itr[subgroup_offset + linear_tid + (idx * GROUP_WORK_ITEMS)]);
+      }
+    }
+ }
+
+private:
+};
+
+ __dpct_inline__ void load(size_t GROUP_WORK_ITEMS,
+                           size_t ITEMS_PER_WORK_ITEM,
+                           load_algorithm ALGORITHM,
+                           InputT (&items)[ITEMS_PER_WORK_ITEM],
+                           InputIteratorT block_itr,
+                           Item &item){
+
+   if (ALGORITHM == blocked){
```

Comment:
These `if`, `else if` will need to be `constexpr` to avoid a runtime branch for each load call.